### DISCUSSION TITLE
DIfference between `fsrs.js` and `ts-fsrs`

### DISCUSSION THREAD

**Discussion by 1cedsoda** [open-spaced-repetition/free-spaced-repetition-scheduler#11] [General]
What is the difference between `fsrs.js` and `ts-fsrs`, since they are both implemented using typescript?

**Comment by L-M-Sherlock** [comment]
They are developed by two different developers. The ts-fsrs has more documents.

**Comment by ishiko732** [comment]
I am one of the authors of `ts-fsrs`. `ts-fsrs` was initially separated from business code, which might make it more suitable for writing projects related to business logic. `ts-fsrs` does not prefer you to use `new FSRS()`. Instead, it prefers you to use `fsrs()` to create instances. `fsrs.repeat/rollback/forget` allow you to mix Date/number/string types of time.


## Difference between `fsrs.js` and `ts-fsrs`

### Language and Refactoring History
  - `ts-fsrs` was written in TypeScript from the beginning.
  - `fsrs.js` was originally written in `JavaScript` and later refactored using `TypeScript`.

### FSRS Versions
  - `ts-fsrs` use [FSRS-4.5 Algorithm](https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#fsrs-45)
  - `fsrs.js` use [FSRS-v4 Algorithm](https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Algorithm#fsrs-v4)

### Functional Differences
 - `ts-fsrs` has two additional application features compared to `fsrs.js`:
   - [rollback](https://github.com/open-spaced-repetition/ts-fsrs/blob/b1efd6cb781a60d4183b7cc75dc2120c87ab5ae8/src/fsrs/fsrs.ts#L104-L140)
    - [forget](https://github.com/open-spaced-repetition/ts-fsrs/blob/b1efd6cb781a60d4183b7cc75dc2120c87ab5ae8/src/fsrs/fsrs.ts#L142-L176)

### Revlog Differences
- The `revlog` in `ts-fsrs` provides more detailed information to support operations such as `rollback` and `forget`.

### Rating Differences
- `ts-fsrs` includes an additional `Manual` Rating option.

### Test Coverage
- `ts-fsrs` maintains a test coverage of over 90%.

**Comment by 1cedsoda** [comment]
Thanks!
